substrate/pkg/controller/substrate/cluster/config.go

```diff
@@ -0,0 +1,305 @@
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/awslabs/kit/substrate/pkg/apis/v1alpha1"
+	"github.com/awslabs/kit/substrate/pkg/utils/discovery"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	certsphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/certs"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/controlplane"
+	etcdphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/etcd"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig"
+	configutil "k8s.io/kubernetes/cmd/kubeadm/app/util/config"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	clusterCertsBasePath = "/tmp/"
+	kubeconfigPath       = "/etc/kubernetes"
+	certPKIPath          = "/etc/kubernetes/pki"
+	clusterManifestPath  = "/etc/kubernetes/manifests"
+	kubeletSystemdPath   = "/etc/systemd/system"
+	kubeletConfigPath    = "/var/lib/kubelet/"
+	kubernetesVersionTag = "v1.21.2-eks-1-21-4"
+	imageRepository      = "public.ecr.aws/eks-distro/kubernetes"
+	etcdVersionTag       = "v3.4.16-eks-1-21-7"
+	etcdImageRepository  = "public.ecr.aws/eks-distro/etcd-io"
+)
+
+type Config struct {
+	S3         *s3.S3
+	S3Uploader *s3manager.Uploader
+}
+
+func (c *Config) Create(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	if substrate.Status.Cluster.Address == nil {
+		return reconcile.Result{Requeue: true}, nil
+	}
+	// ensure S3 bucket
+	if err := c.ensureBucket(ctx, substrate); err != nil {
+		return reconcile.Result{}, fmt.Errorf("ensuring S3 bucket, %w", err)
+	}
+	// create all configs file
+	cfg := defaultClusterConfig(substrate)
+	if err := c.generateCerts(cfg, substrate); err != nil {
+		return reconcile.Result{}, fmt.Errorf("generating certs, %w", err)
+	}
+	if err := c.kubeConfigs(cfg, substrate); err != nil {
+		return reconcile.Result{}, fmt.Errorf("generating kube config, %w", err)
+	}
+	if err := c.generateStaticPodManifests(cfg, substrate); err != nil {
+		return reconcile.Result{}, fmt.Errorf("generating manifests, %w", err)
+	}
+	if err := c.kubeletSystemService(cfg, substrate); err != nil {
+		return reconcile.Result{}, fmt.Errorf("generating kubelet service config, %w", err)
+	}
+	// upload to s3 bucket
+	if err := c.S3Uploader.UploadWithIterator(ctx, NewDirectoryIterator(
+		aws.StringValue(discovery.Name(substrate)), path.Join(clusterCertsBasePath, aws.StringValue(discovery.Name(substrate))))); err != nil {
+		return reconcile.Result{}, fmt.Errorf("uploading to S3 %w", err)
+	}
+	logging.FromContext(ctx).Infof("Uploaded cluster configuration to s3://%s", aws.StringValue(discovery.Name(substrate)))
+	return reconcile.Result{}, nil
+}
+
+func (c *Config) Delete(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	// delete the s3 bucket
+	if err := s3manager.NewBatchDeleteWithClient(c.S3).Delete(ctx, s3manager.NewDeleteListIterator(
```

Comment:
I don't think you need to delete the files. Just delete the bucket.

Reply:
It fails to delete the bucket if it still has files in it, I will check if there is a `--force` kind of option

Reply:
No!?

Reply:
Looked at this there is no force option in the request